### PR TITLE
ACC 288 improve error codes

### PR DESCRIPTION
**Pull request by philippemiguet**
Heya @hharnisc (cc @esclapes @CoxyBoris), I've started using your latest work with buffer-rpc, but I've got a bunch of questions that are slowing me down a bit. I would love your early thoughts regarding how we are about to use buffer-rpc errors, and how I'm approaching things here :)

**Questions**
First of all, a few questions:
 - What is the meaning of `handled:true` and `handled:false`? Does `handle:true` means an expected error and `handled:false` an unexpected error?
 - What is the meaning of `rpcError:true`?
 - What is the statusCode for? Can we set it to 404 when an account is not found? Should we set it to 500 when an unexpected error occurs?

**Context**
The main goal of this PR is to provide error codes for each expected error. For instance, when a Reply user will log in for the first time, we will want to know if their email already exists in the core-auth-svc using `getAccount`, if not we want to create an account and product link for them.

If we don't have error code to make sense of the error returned from `getAccount`, we don't know if the error we got is because the email was not found and that we should create an account, or if the error was something else and that we should skip and display an error to the user.

I left many in-line comments to help you understand what I did so far!

**Questions regarding this PR**
What I tried to do with this WIP is to return custom error codes for expected errors in the RPC handler. Then came new questions:
 - What to do with expected errors from the authenticationModel, like "the password is too short" error? Should it be a handled error? It feel so!
 - What to do with unexpected errors from the authenticationModel, like MongoDB errors due to indexes rules, or MongoDB being down, or a syntax error?
 - How can we make sure those two types of error are returned properly?
 - How does all this play with the errorMiddleware?

### CODE REVIEW

File: src/libs/errors.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+    GENERIC_ERROR: "6000",
+    REQUIRED_PARAMETER_MISSING: "6001",
+    INVALID_PARAMETER: "6002",
+    ACCOUNT_NOT_FOUND: "6003",
+}
```
Comment (philippemiguet) [inline src/libs/errors.js]: We took 6000 as a random range, curious to hear if you have any recommendations regarding how different services will use error codes. Should we prefix them by the service name, i.e `core_auth_6000`?

File: src/models/authenticationAccountModel.js [hunk 2]
```diff
@@ -90,9 +92,10 @@ authenticationAccountSchema.methods.setPassword = async function({ password }) {
   if (!password) throw new Error('Password is required')
   if (typeof password !== 'string') throw new Error('Password must be a String')
   if (password.length < MIN_PASSWORD_LENGTH)
-    throw new Error(
-      `Password must be at least ${MIN_PASSWORD_LENGTH} characters`,
-    )
+    throw createError({
+      message: `Password must be at least ${MIN_PASSWORD_LENGTH} characters`,
+      code: INVALID_PARAMETER,
+    })
```
Comment (philippemiguet) [inline src/models/authenticationAccountModel.js]: How does it feel to use `createError()` here, instead of `new Error()`. By using createError(), we will be able to tell in the global try-catch of the RPC handlers the error is a handled/expected error. While if we simply throw a new error, this error will be caught and handled in the same way as an unexpected error and may end up in Bugsnag.

note: I only changed that one line for now just to kickoff the discussion.

File: src/rpc/createAccount.js [hunk 3]
```diff
@@ -1,29 +1,49 @@
...
+    if (!error.handled) {
+      error = createError({
+        message: error.message,
+        code: GENERIC_ERROR,
+        handled: false,
+        statusCode: 500
+      })
+    }
+
+    throw error
```
Comment (philippemiguet) [inline src/rpc/createAccount.js]: Boris suggested to wrap the handler in a global try-catch, then if all expected errors are created using the createError() function with `{handled: true}`, here we would be able to catch unexpected errors and return a specific error code for them.

The goal is for the login and reply services to be able to understand the errors we get from the core-auth-svc and know which one to ignore vs which one to show to the users: any expected/handled errors should be hidden (i.e. invalid password or email when calling verifyAccount endpoint), while unexpected errors should be shown while hiding the details (i.e. service is down, we may want the user to know something is not working).

I know the errorMiddleware is also taking care of unexpected errors, curious to hear if this approach may interfere with how you wanted the errorMiddleware to work?

File: src/rpc/getAccount.js [hunk 4]
```diff
@@ -1,22 +1,47 @@
...
+    if (!error.handled) {
+      error = createError({
+        message: error.message,
+        code: GENERIC_ERROR,
+        handled: false,
+        statusCode: 500
+      })
+    }
+
+    throw error
```
Comment (philippemiguet) [inline src/rpc/getAccount.js]: Same as the other endpoint :)

**Comment by hharnisc** [comment]
Hey @philippemiguet 

**handled**

I've written some more documentation on when to use the handled flag: https://github.com/bufferapp/buffer-rpc/blob/master/README.md#handled-flag

TL;DR is to leave it at the default (handled = true) when using `createError` unless you've got a good reason to change it.

**rpcError:true**

This is a flag internal to `buffer-rpc` and is only set when you use `createError`. It indicates that the error came from a `createError` function, if it didn't come from `createError` (say thrown by a third party software) it passes the error off the error handling middleware.

https://github.com/bufferapp/buffer-rpc/blob/master/src/rpc.js#L41

**statusCode**

Within the RPC context:

**404** RPC endpoint could not be found
**500** unexpected error occured
**400** an error was triggered by createError
**200** success
**XXX** any other code doesn't mean anything inside the context of RPC, you can still set this if you need the client to do something

I'll leave the rest of the comments inline :smile:

File: src/libs/errors.js (hunk 1 above)
Comment (hharnisc) [reply src/libs/errors.js]: At this point these error codes don't mean anything outside of the context of the `Authentication Service`. At some point we might want to consolidate errors into a repo, but we don't have that issue yet. The error codes themselves should probably be integers here too.

File: src/models/authenticationAccountModel.js (hunk 2 above)
Comment (hharnisc) [reply src/models/authenticationAccountModel.js]: I'd strongly recommend not using the global try/catch here since Buffer RPC is already doing that: https://github.com/bufferapp/buffer-rpc/blob/master/src/rpc.js#L24-L38

`throw createError` should be enough here -- Buffer RPC would return a response that looks like this:

```
// statusCode = 400
{
  error: 'Password must be at least XXX characters',
  code: 6002,
  handled: true,
}
```

The error here should be handled because there's no partial error that occurred (only the user need to retry with different password).

File: src/rpc/getAccount.js (hunk 4 above)
Comment (hharnisc) [reply src/rpc/getAccount.js]: (same as above)

File: src/rpc/createAccount.js (hunk 3 above)
Comment (hharnisc) [reply src/rpc/createAccount.js]: (from above) I'd avoid the redundant global try/catch. Thow the error in line when it happens with the expected error code -- then you're done :smile:

File: src/libs/errors.js (hunk 1 above)
Comment (philippemiguet) [reply src/libs/errors.js]: I see one use case where it might be useful to know if the error belongs to one or another service. In the login service we may do:

```
try {
 // await call to core-auth-svc
 //await call to reply-internal-api
} catch(error) {
  //here we don't know the source of the error
}
```

The way to do it right now is to wrap each call in a separate `try-catch` :)

File: src/models/authenticationAccountModel.js (hunk 2 above)
Comment (philippemiguet) [reply src/models/authenticationAccountModel.js]: Oh, I see it makes sense!

I think we were also missing using the errorMiddleware after the rpc handler for the POST `/rpc` route: this was causing to trigger a fetch error on unhandled errors. This is what also led us to add the unecessary global try-catch :)

Output of unhandled error without errorMiddleware
```
login_1                              | { FetchError: invalid json response body at http://core-authentication-service:3000/rpc reason: Unexpected token < in JSON at position 0
login_1                              |     at /usr/src/app/node_modules/node-fetch/lib/body.js:48:31
login_1                              |     at <anonymous>
login_1                              |     at process._tickCallback (internal/process/next_tick.js:160:7)
login_1                              |   name: 'FetchError',
login_1                              |   message: 'invalid json response body at http://core-authentication-service:3000/rpc reason: Unexpected token < in JSON at position 0',
login_1                              |   type: 'invalid-json' }
login_1                              | message: invalid json response body at http://core-authentication-service:3000/rpc reason: Unexpected token < in JSON at position 0
login_1                              | code: undefined
login_1                              | status: undefined
```

Output of unhandled error with errorMiddleware
```
login_1                              | { Error: E11000 duplicate key error collection: buffer.accounts index: email_1 dup key: { : "phil@test.fr" }
login_1                              |     at fetch.then.then (/usr/src/app/node_modules/micro-rpc-client/index.js:42:23)
login_1                              |     at <anonymous>
login_1                              |     at process._tickCallback (internal/process/next_tick.js:160:7) code: 5000, handled: false, status: 500 }
login_1                              | message: E11000 duplicate key error collection: buffer.accounts index: email_1 dup key: { : "phil@test.fr" }
login_1                              | code: 5000
login_1                              | status: 500
```

**tl;dr: I will add the errorMiddleware and remove the global try-catch!**

@hharnisc can you confirm using the errorMiddleware is the right thing to do?

File: src/index.js [hunk 5]
```diff
@@ -70,6 +71,7 @@ const main = async () => {
       method('completePasswordReset', completePasswordReset),
       method('verifyAccount', verifyAccount),
     ),
+    errorMiddleware
```
Comment (philippemiguet) [inline src/index.js]: @hharnisc - as shared on the other comment, I think we were missing this middleware here to unsure unhandled error will be returned as a proper error and won't trigger a fetch error!

Should all our RPC services use the errorMiddleware? Now I've played with it, it seems like a good practice 😄

**Comment by philippemiguet** [comment]
Hey @hharnisc thanks for the extra context regarding all the error attributes, it was really helpful to understand the concepts behind those :) 

I've left another set of comments for you, if you have time to review them :)

File: src/rpc/verifyAccount.js [hunk 6]
```diff
@@ -1,41 +1,65 @@
...
+
   if (
-    !user ||
-    !(await user.verifyPassword({ password })) ||
-    !(await user.verifyProductName({ productName }))
+    !account ||
+    !(await account.verifyPassword({ password })) ||
+    !(await account.verifyProductName({ productName }))
   ) {
     // intentionally generic error
```
Comment (philippemiguet) [inline src/rpc/verifyAccount.js]: Hey @hharnisc, curious to hear why we want to return generic errors from the core-auth-svc? Being only an internal service, is it still needed to hide some of our logic?

**Comment by philippemiguet** [comment]
Hey @hharnisc and @CoxyBoris this is ready for a review 👍

File: src/rpc/verifyAccount.js (hunk 6 above)
Comment (hharnisc) [reply src/rpc/verifyAccount.js]: There are a couple reasons for this, the first is that when we initially built this service there was a possibility that it was going to be exposed to the public internet. The second is that it reduces the chance of someone phishing for email addresses. If we absolutely need to know the failure type, we could add it.

File: src/models/authenticationAccountModel.js (hunk 2 above)
Comment (hharnisc) [reply src/models/authenticationAccountModel.js]: This looks correct to me @philippemiguet :raised_hands: 

The session service has not been upgraded yet :smile: the only application that has been upgraded is `Publish` (and this one very soon!)

Comment (philippemiguet) [reply src/models/authenticationAccountModel.js]: Good to know, thanks :D

**Comment by philippemiguet** [comment]
Thanks @hharnisc for all the help there! I've fixed the tests, totally missed they were failing: since I changed a lot of error messages it was to be expected though 🤦‍♂️
